packages/js/src/cache/notifications-cache.ts

```diff
@@ -0,0 +1,212 @@
+  'notification.read.pending',
+  'notification.read.resolved',
+  'notification.unread.pending',
+  'notification.unread.resolved',
+  'notification.archive.pending',
+  'notification.archive.resolved',
+  'notification.unarchive.pending',
+  'notification.unarchive.resolved',
+  'notification.complete_action.pending',
+  'notification.complete_action.resolved',
+  'notification.revert_action.pending',
+  'notification.revert_action.resolved',
+];
+
+const multipleNotificationEvents: NotificationsEvents[] = [
+  'notifications.read_all.pending',
+  'notifications.read_all.resolved',
+  'notifications.archive_all.pending',
+  'notifications.archive_all.resolved',
+  'notifications.archive_all_read.pending',
+  'notifications.archive_all_read.resolved',
+];
+
+export class NotificationsCache {
+  #emitter: NovuEventEmitter;
+  /**
+   * The key is the stringified notifications filter, the value is the pagination cache
+   * where the key is pagination arguments and the value is the list of notifications.
+   */
+  #cache: Cache<Cache<ListNotificationsResponse>>;
+  /**
+   * The key is the stringified notifications filter, the value is the set of unique notification ids.
+   */
+  #idsCache: Cache<Set<string>>;
+
+  constructor() {
+    this.#emitter = NovuEventEmitter.getInstance();
+    singleNotificationEvents.forEach((event) => {
+      this.#emitter.on(event, this.handleSingleNotificationEvent);
+    });
+    multipleNotificationEvents.forEach((event) => {
+      this.#emitter.on(event, this.handleMultipleNotificationsEvent);
+    });
+    this.#cache = new InMemoryCache();
+    this.#idsCache = new InMemoryCache();
+  }
+
+  private getFilterCache = (args: ListNotificationsArgs): Cache<ListNotificationsResponse> | undefined => {
+    return this.#cache.get(getFilterKey(args));
+  };
+
+  private updateNotificationInCache = (cache: Cache<ListNotificationsResponse>, data: Notification): void => {
+    const allCache = cache.pairs();
+
+    allCache.forEach(([key, el]) => {
+      const index = el.notifications.findIndex((notification) => notification.id === data.id);
+      if (index === -1) {
+        return;
+      }
+
+      const updatedNotifications = [...el.notifications];
+      updatedNotifications[index] = data;
+
+      cache.set(key, { ...el, notifications: updatedNotifications });
+    });
+  };
+
+  private handleSingleNotificationEvent = ({ data }: { data?: Notification }): void => {
+    if (!data) {
+      return;
+    }
+
+    const filterKeysToUpdate = this.#idsCache.keys().filter((key) => {
+      const cache = this.#idsCache.get(key);
+
+      return cache?.has(data.id);
+    });
+
+    filterKeysToUpdate.forEach((key) => {
+      const filterCache = this.#cache.get(key);
+      if (!filterCache) {
+        return;
+      }
+
+      this.updateNotificationInCache(filterCache, data);
+
+      const notificationsResponse = this.getReducedNotifications(filterCache);
+      this.#emitter.emit('notifications.list.updated', {
```

Comment:
This is the event that the UI is listening to. Of course, the name is a matter of discussion.